### ISSUE TITLE
Running over tunneled sockets

### ISSUE THREAD

**Issue by afbjorklund** [containerd/nerdctl#131]
Is it possible to run the `nerdctl` client over tunneled sockets ?

Trying to run, but getting errors from the supposed local rootless.

```console
$ ctr --version
ctr containerd.io 1.4.4
$ echo $CONTAINERD_ADDRESS 
./containerd.sock
```

```console
$ ctr version
Client:
  Version:  1.4.4
  Revision: 05f951a3781f4f2c1911b05e61c160e9c30eaa8e
  Go version: go1.13.15

Server:
  Version:  1.4.3
  Revision: 269548fa27e0089a8b8278fc4fc781d7f65a939b
  UUID: dbc0f319-4fee-4dc1-9c5f-1e3433fe6c68
WARNING: version mismatch
WARNING: revision mismatch
```

```console
$ nerdctl --version
nerdctl version 0.7.1
$ nerdctl version
FATA[0000] rootless containerd not running? (hint: use `containerd-rootless-setuptool.sh install` to start rootless containerd): stat /run/user/1000/containerd-rootless: no such file or directory 
```

~~Similar with BuildKit, but couldn't find the environment variable for it.~~

It also has a _slightly_ different syntax on the param and the address.

```console
$ buildctl debug workers
ID				PLATFORMS
m54eigt1xxnto7i8iihho63dw	linux/amd64
$ echo $BUILDKIT_HOST 
unix://buildkitd.sock
```

----

EDIT: Did find BUILDKIT_HOST eventually, didn't look hard enough.

Guess it rhymes with DOCKER_HOST and not CONTAINERD_ADDRESS

**Comment by afbjorklund** [comment]
Trying to run as root gives a different error:

`FATA[0000] failed to dial "./containerd.sock": failed to build resolver: invalid (non-empty) authority: . `

**Comment by afbjorklund** [comment]
Did find BUILDKIT_HOST eventually, didn't look hard enough.

**Comment by AkihiroSuda** [comment]
`ssh -L` socket cannot be supported, because nerdctl and containerd need to share the same filesystem.

So just please use `ssh user@host nerdctl ...` .

If you want to expose your local filesystem to the remote ssh host, https://github.com/AkihiroSuda/sshocker might be useful:
`sshocker -v $(pwd):/mnt user@host nerdctl build -t foo /mnt` (I haven't tried, but I guess it should work).

Also, I think I recently saw something similar on Hackernews, but can't remember its name.

**Comment by afbjorklund** [comment]
The main use case was building, but I have a workaround wrapper for it already. Basically it is using scp for the files...

I think removing Docker from Kubernetes was a bit premature, and replacing it with ctr/buildctl/nerdctl isn't very friendly.

**Comment by AkihiroSuda** [comment]
FYI BuildKit can be executed rootless and on Kubernetes: https://github.com/moby/buildkit/tree/master/examples/kubernetes

**Comment by afbjorklund** [comment]
> FYI BuildKit can be executed rootless and on Kubernetes

That is great! (I think we saw similar with `kubectl build`)

The main feature that we were looking for was building directly to the storage of the cluster, to avoid pushing and pulling.

Previously we used Docker to do this (when it was the container runtime), but now using BuildKit with containerd worker.

The only problem was the client, but has been worked around.
